### PR TITLE
distance approximation functions

### PR DESCRIPTION
**Pull request by NutchaW**
probably need more tests later. Please let me know if there are issues.

### CODE REVIEW

File: R/approx_distance.R [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+#' for a single target, location, target_end_date
+#' @param approx_rule a string specifying which formula to use
+#' for approximation. Valid rules are "left-sided" and "trapezoid".
+#' The default rule is "trapezoid".
+#' @return a data frame with approximated pairwise Cramér distances
+#' for all possible pairs of models
+#'
+#' @export
+#'
+cd_combination <- function(single_frame, end_date, approx_rule) {
```
Comment (nickreich) [inline R/approx_distance.R:130]: I suggest a different verb than "combination". perhaps both functions should have more active and descriptive names. How about:
`calc_one_cramers_dist()` for what is now `cd_pairwise()`
`calc_multiple_cramers_dist()` for what is now `cd_combination()`

File: R/approx_distance.R [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+  if (approx_rule == "left-sided" |
+      approx_rule == "trapezoid") {
+    tau_F <- tau_G <- unique(single_frame$quantile)
+  } else {
+    tau_F <- tau_G <- NULL
+  }
+  # remove any models with NA values
+  single_tarloc <-
+    single_frame[, colSums(is.na(single_frame)) == 0] %>%
```
Comment (nickreich) [inline R/approx_distance.R:139]: I think if you remove any models, you should throw a warning/message saying "The following model(s) were removed due to having NAs in some quantile values."

File: R/approx_distance.R [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+#'
+#' @export
+#'
+cd_combination <- function(single_frame, end_date, approx_rule) {
+  if (approx_rule == "left-sided" |
+      approx_rule == "trapezoid") {
+    tau_F <- tau_G <- unique(single_frame$quantile)
+  } else {
+    tau_F <- tau_G <- NULL
+  }
```
Comment (nickreich) [inline R/approx_distance.R:136]: maybe also need to check that there is just one location and/or target_end_date?

File: R/approx_distance.R [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+  # remove any models with NA values
+  single_tarloc <-
+    single_frame[, colSums(is.na(single_frame)) == 0] %>%
+    dplyr::filter(target_end_date == end_date) %>%
```
Comment (nickreich) [inline R/approx_distance.R:140]: end_date is not defined in the @param documentation. is the ultimate goal to have this function take a data_frame that has potentially multiple locations, horizons, and dates and have it average over 0,1 or 2 of those dimensions? I suggest that the desired default behavior is to average across all locations, horizons, and dates provided. If we want to enforce for now that only 1 location/1 date are provided we can do that, and then extend later to average over multiple. or we could build in support now to average over multiple locations/dates/horizons.

This is a complicated piece, and we could pull this out of the PR for now to give us time to sort it out later.

File: R/approx_distance.R [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+      q0 <- c(q_F, q_G)
+      a0 <- c(rep(1, N), rep(0, M))
+      # indicator whether the entry is from F or G
+      a <- a0[order(q0)]
+      q <- q0[order(q0)]
+      tau <- c(tau_F, tau_G)[order(q0)]
+      diffs_q <- diff(q)
+      # prob level vectors
+      tau_F_v <- a * tau
+      tau_F_v2 <- sapply(1:(N + M),
```
Comment (nickreich) [inline R/approx_distance.R:82]: if ok, I'm going to leave it to @aaronger to check the math on this.

File: R/approx_distance.R [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+      }
+      metric <- cvm
+    }
+    return(metric)
+  }
+
+#' Approximate pairwise Cramér distance for a single target,
+#' date, location data frame
+#'
+#' @param single_frame a data frame containing correctly-formatted forecasts
```
Comment (nickreich) [inline R/approx_distance.R:120]: I suggest calling this something more descriptive. how about `forecast_data`? and indicating in the documentation that this is "a data frame containing forecasts formatted as returned from load_forecasts()` so you can link to the specification there?

File: R/approx_distance.R (hunk 3 above)
Comment (nickreich) [reply R/approx_distance.R:136]: and target variable? and horizon?

File: R/approx_distance.R [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+  # pairwise column calculation
+  tmp <- single_tarloc %>%
+    dplyr::select(
+      -c(
+        "target_variable",
+        "target_end_date",
+        "location",
+        "type",
+        "quantile",
+        "horizon",
```
Comment (nickreich) [inline R/approx_distance.R:151]: Are multiple horizons getting mushed together here? it's not clear to me how this is working.

File: R/approx_distance.R [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+#' for multiple targets, locations, target_end_dates
+#' @param approx_rule a string specifying which formula to use
+#' for approximation. Valid rules are "left-sided" and "trapezoid".
+#' The default rule is "trapezoid".
+#' @return a data frame with approximated pairwise Cramér distances
+#' for all possible pairs of models
+#'
+#' @export
+#'
+approx_distance <- function(forecasts, approx_rule) {
```
Comment (nickreich) [inline R/approx_distance.R:190]: I'm not reviewing this function in detail.

File: R/pairwise_filter.R [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+#' @param dataframe required data.frame with forecasts in the format returned
+#' by load_forecasts
+#'
+#' @return data frame with columns model, location, horizon,
+#' target_variable, target_end_date, type, quantile, value,
+#' abbreviation
+#'
+#' @export
+#'
+pairwise_filter <- function(dataframe) {
```
Comment (nickreich) [inline R/pairwise_filter.R:12]: suggest removing this for now.

**Review by nickreich — CHANGES_REQUESTED** [review]
This is a great start at getting these ideas down into code. My suggestion is to make this first PR much smaller. Let's focus for starters on getting the foundational cramers distance computation in a good, tested format, without worrying about supporting any aggregation for starters. So my recommendation is to remove all of the aggregated computation functions and just focus on the single pairwise function for starters. I have some specific suggestions for argument validation, naming, and possible breaking into separate functions the specific computations in the comments of the review.

Additionally, I think an essential piece for this PR is a few simple unit tests with a few very simple vectors of quantiles. For example, write a test function that defines manually a simple q_F and q_G where we know what the cramer's distance calculation should be exactly. Then, apply the functions and make sure the right answer pops out. Maybe define a few with incorrect quantile vectors (e.g. what if the quantiles aren't in order? or what if the specified taus don't match in length or `length(tau_F) != length(q_F)`) to check that things error correctly.

File: R/approx_distance.R [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+      # compute the lengths of segments defined by sorted quantiles:
+      diffs_q <-
+        c(diff(q), 0) # zero necessary for indexing below, but we could put
+      # anything (gets multiplied w zero)
+      
+      # and approximate CD
+      if (approx_rule == "equal_1") {
+        cvm <- sum(diffs_q * b * (b + 1)) / (K + 1) / (K)
+      } else if (approx_rule == "equal_2") {
+        cvm <- sum(diffs_q * b ^ 2 / K ^ 2)
```
[the remark this answers is not in the document]
Comment (aaronger) [reply R/approx_distance.R:64]: Also maybe we should rely on R's associativity here and just write `/ ((K+1)*K)` in line 62

File: R/pairwise_cd_equal.R [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+#' of `q_F` and `q_G` and \eqn{b_i} is an element of a vector of the absolute
+#' values of cumulative sums of \eqn{\mathbf{a}}, whose element is 1 if
+#' \eqn{q_i} is a quantile of F or -1 if \eqn{q_i} is a quantile of G.
+#' @example
+#' f_vector <- 1:10
+#' g_vector <- seq(2,20,2)
+#' pairwise_cd_equal(f_vector,g_vector,"second_rule")
+#' @export
+#'
+pairwise_cd_equal <- function(q_F, q_G, approx_rule) {
```
Comment (nickreich) [inline R/pairwise_cd_equal.R:28]: I understand that it doesn't matter what the probability levels are, but maybe we should still ask for them? even if just a single tau argument that would apply to both q_F and q_G?

Comment (nickreich) [reply R/pairwise_cd_equal.R:28]: Seems like it is something we should still be thinking about validating.

Comment (NutchaW) [reply R/pairwise_cd_equal.R:28]: We could, but that information will not be used at all anywhere in the function. Maybe we will have a use for that in an aggregation function (and can ask for it then)?

Comment (NutchaW) [reply R/pairwise_cd_equal.R:28]: validating = making sure the quantiles from the two forecasts correspond to the same probability levels? If so, then we can add tau_F and tau_G arguments

File: tests/testthat/test-pairwise_cd_equal.R [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+  f_vector2 <- qunif(tau2,0,1)
+  g_vector2 <- qunif(tau2,0,2)
+  approx_cd2 <- pairwise_cd_equal(f_vector2,g_vector2, "first_rule")
+  # check that the approx. cd of large k is closer to true cd
+  diff1 <- abs(true_cd - approx_cd1)
+  diff2 <- abs(true_cd - approx_cd2)
+  expect_lte(diff2,diff1)
+})
+
+test_that("second_rule approximation is reasonalble", {
```
Comment (nickreich) [inline tests/testthat/test-pairwise_cd_equal.R:69]: typo: reasonable

Comment (nickreich) [reply tests/testthat/test-pairwise_cd_equal.R:69]: several times in the file.

**Review by nickreich — CHANGES_REQUESTED** [review]
the new tests look really great!

A few specific and overall changes requested:
 - please add additional authorship (Nutcha) and appropriate info in DESCRIPTION and NEWS.md. In general, follow contribution guidelines here: https://github.com/reichlab/covidHubUtils/blob/master/README.md#contributing-guidelines
  - my suggestion from the first review was to have a single function that would then call each the appropriate version of each of these two cd functions. This will increase usability from the user perspective. So the current PR needs a wrapper function with an accessible name [I suggest `calc_cramers_dist()`] that calls the appropriate helper function for the actual computation. I documented some of these ideas separately in a spec sheet.

**Review by aaronger — COMMENTED** [review]
Submitting this now to head-off confusion from possible updates tonight or tomorrow.  In the meantime, I will try to write some tests for more complicated distributions, e.g., triangular.

File: R/calc_cramers_dist_equal_space.R [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+        isTRUE(all.equal((range(diff(tau_G_ordered))/mean(diff(tau_G_ordered)))[1], 
+                          (range(diff(tau_G_ordered))/mean(diff(tau_G_ordered)))[2],.Machine$double.eps ^ 0.5))==FALSE) {
+      warning(
+        "tau_F_ordered and tau_G_ordered are not equaly-spaced, this approximation rule is based on equally-spaced probability levels"
+      )
+    }
+    # run select rule
+    # compute quantile levels from length of provided quantile vectors:
+    K <- length(q_F_ordered)
+    p <-
```
Comment (nickreich) [inline R/calc_cramers_dist_equal_space.R:91]: p is not used later in this function.

**Review by nickreich — CHANGES_REQUESTED** [review]
Looking really really good. I have just a few minor comments that I don't think need to be addressed prior to pulling and could be added as issues to address in a future version. 

Should we add a check/test that values in the ordered q_F and q_G vectors are increasing? 

@NutchaW - you should add yourself as a contributor to the authors field in the DESCRIPTION file!

File: R/calc_cramers_dist_one_model_pair.R [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+#' where \eqn{q_i} is an element in a vector of an ordered pooled quantiles
+#' of `q_F` and `q_G` and \eqn{\tau^F_j} and \eqn{\tau^G_j} are defined as
+#' the probability level of a quantile in `q_F` when \eqn{q_i} comes from \eqn{F} and
+#' the probability level of a quantile in `q_G` when \eqn{q_i} comes from \eqn{G},
+#' respectively.
+#' @example
+#' f_vector <- 1:9
+#' tau_F_vector <- tau_G_vector <- seq(0.1,0.9,0.1)
+#' g_vector <- seq(4,20,2)
+#' calc_cramers_dist_one_model_pair(f_vector,tauF_vector,g_vector,tauG_vector,"left-sided")
```
Comment (nickreich) [inline R/calc_cramers_dist_one_model_pair.R:49]: the approx rule is invalid here. Should be "left_sided_riemann"?

**Review by nickreich — CHANGES_REQUESTED** [review]
the PR should have an update to the NAMESPACE file, since you are adding exported functions. Have you run a "Clean and Rebuild" in RStudio? If you have Roxygen building turned on this should add the necessary lines for you.

File: tests/testthat/test-pairwise_cd_equal.R (hunk 12 above)
Comment (NutchaW) [reply tests/testthat/test-pairwise_cd_equal.R]: the power of copy and paste

**Comment by NutchaW** [comment]
Updated PR is here. About check for increase quantiles, I forced the sorting to happen (with a warning when it does) if the inputs are not in the right order.

**Pull request merged**